lambda/value_rec_compiler.ml

```diff
@@ -106,80 +123,153 @@ let join_sizes size1 size2 =
   | Unreachable, size | size, Unreachable -> size
   | _, _ -> dynamic_size ()
 
-let compute_static_size lam =
-  let rec compute_expression_size env lam =
+let compute_static_size bound_id lam =
+  let rec compute_expression_size lam =
     match lam with
-    | Lvar v ->
-      begin match Ident.Map.find_opt v env with
-      | None ->
-        dynamic_size ()
-      | Some binding_size ->
-        Lazy_backtrack.force
-          (fun { lambda; env } -> compute_expression_size env lambda)
-          binding_size
-      end
+    | Lvar id -> lam, Variable { id; need_update_dummy = true }
     | Lmutvar _ -> dynamic_size ()
-    | Lconst _ -> Constant
+    | Lconst _ -> lam, Constant
     | Lapply _ -> dynamic_size ()
-    | Lfunction _ -> Function
-    | Llet (_, _, id, def, body) ->
-      let env =
-        Ident.Map.add id (Lazy_backtrack.create { lambda = def; env }) env
-      in
-      compute_expression_size env body
-    | Lmutlet(_, _, _, body) ->
-      compute_expression_size env body
+    | Lfunction _ -> lam, Function
+    | Llet (kind, vk, id, def, body) ->
+      let body, size = compute_expression_size body in
+      begin match size with
+      | Variable { id = id'; need_update_dummy } when Ident.same id id' ->
+        let new_def, size = compute_expression_size def in
+        let def = if need_update_dummy then new_def else def in
+        Llet(kind, vk, id, def, body), size
+      | _ ->
+        Llet(kind, vk, id, def, body), size
+      end
+    | Lmutlet(vk, id, def, body) ->
+      let body, size = compute_expression_size body in
+      Lmutlet(vk, id, def, body), size
     | Lletrec (bindings, body) ->
-      let env =
-        List.fold_left (fun env_acc { id; def = _ } ->
-            Ident.Map.add id (Lazy_backtrack.create_forced Function) env_acc)
-          env bindings
-      in
-      compute_expression_size env body
+      let body, size = compute_expression_size body in
+      begin match size with
+      | Variable { id = id'; need_update_dummy = _ }
+        when List.exists (fun { id; _ } -> Ident.same id id') bindings ->
+        Lletrec (bindings, body), Function
+      | _ ->
+        Lletrec (bindings, body), size
+      end
     | Lprim (p, args, _) ->
-      size_of_primitive env p args
-    | Lswitch (_, sw, _) ->
+      begin match size_of_primitive p args with
+      | (Constant | Function | Unreachable) as size ->
+        lam, size
+      | Block _ as size ->
+        add_update_dummy bound_id lam, size
+      | Variable { id; need_update_dummy = _ } ->
+        (* This corresponds to an allocation of a yet unknown size.
+           We can still add the [caml_update_dummy] call, but must
+           make sure that it doesn't get duplicated.
+           Note that if the size ends up not being a Block, we will need to
+           drop this rewriting. *)
+        add_update_dummy bound_id lam,
+        Variable { id; need_update_dummy = false }
```

Comment:
I find it difficult to rebase this, because this sorts of assume that there is a single `update_dummy` primitive that works for all sizes, which is not the design anymore with trunk. I would need to know in advance whether a variable is going to need an `update_dummy_lazy` or not.

But I have a question about this. Can `size_of_primitive` actually return a `Variable` case? I don't understand when that happens and I wonder if this is dead code. (If it is indeed the case, this might suggest a distinction between two different types of sizes, "closed" sizes that do not allow the `Variable` case an "open" sizes that classify open terms with free variables.)